test/tools-test.cpp

```diff
@@ -0,0 +1,196 @@
+#include "gtest/gtest.h"
+
+#include "../src/tools.h"
+#include <magic.h>
+
+std::string directoryPath;
+magic_t magic;
+bool inflateHtmlFlag = false;
+bool isVerbose() { return false; }
+
+TEST(CommonTools, getFileSize)
+{
+  std::string fn = "data/minimal-content/favicon.png";
+
+  system("pwd");
+
+  ASSERT_TRUE(fileExists(fn));
+  EXPECT_EQ(getFileSize(fn), 2725);
+}
+
+TEST(CommonTools, base64_encode)
+{
+  unsigned char data[] = { 0xff, 0x00, 0x7a };
+  std::string txt = base64_encode(data, sizeof(data));
+  EXPECT_EQ(txt, "/wB6");
+}
+
+TEST(CommonTools, decodeUrl)
+{
+  // test normal use
+  const char *src = "https://www.example.com/cgi-bin/search.cgi?q=example%20search";
+  std::string res = decodeUrl(src);
+  EXPECT_EQ(res, "https://www.example.com/cgi-bin/search.cgi?q=example search");
+
+  src = "%2a";
+  res = decodeUrl(src);
+  EXPECT_EQ(res, "*");
+
+  src = "%00";
+  res = decodeUrl(src);
+  EXPECT_EQ(res.size(), 1);
+  EXPECT_EQ(res[0], '\0');
+
+  src = "%ff";
+  res = decodeUrl(src);
+  EXPECT_EQ(res.size(), 1);
+  EXPECT_EQ(res[0], '\xff');
+
+
+  // test corner cases
+  src = "";
```

Comment:
Here we could have a map of (original string, expected string) and then run it on a loop.

_pseudo code_
```
unordered_map<string, string> = expectationsMap = {  {"%", "%"}, {"%2", "%2"}, ..., ... };

for (auto p : expectationsMap) EXCPECT_EQ(p.first, p.second);

```

But its just a suggestion. No need to implement it for approval.

(I later noticed that there are other places in the code that this could be done)

Reply:
Good idea to have at least one example in the code. One test was rewritten.